### ISSUE TITLE
Terraform compile + aws-sdk-go MaxRetries

### ISSUE THREAD

**Issue by toddrosner** [hashicorp/terraform#4530]
Frequently Terraform runs into API request rate limits even if parallelism is set to 0. I see that there are others that have this issue, and one of the mentioned fixes was to compile Terraform from master, then update the MaxRetries logic to 11 in the aws-sdk-go/config.go file here https://github.com/aws/aws-sdk-go/blob/master/aws/config.go#L63

I think it'd be great if there were a way to make this configurable in Terraform so it's a bit more automated, but at least for now this is a potential fix...except I can't seem to get Terraform to compile properly using the directions on the Terraform page here https://github.com/hashicorp/terraform

When running through the compile steps I hit a wall with updating the dependencies and get the following error:

```make: *** [updatedeps] Error 123```

So my question is twofold:

1. Is it yet possible to control the rate at which Terraform makes API calls? The parallelism flag doesn't seem to help...or perhaps there's more to it than the documentation reveals?
2. How can I get around the updatedeps error mentioned above so that I can compile Terraform and set MaxRetries in the aws-sdk-go config.go file?

Any help is much appreciated.

**Comment by jen20** [comment]
Hi @toddrosner! The error with `make updatedeps` may be because you have one of the dependencies already cloned in your `GOPATH`, but left on a branch with no upstream tracking branch. Normally the output of `make` should show which one is causing the problem. I'll have to defer to @catsby on the question about API rate limiting though.

**Comment by toddrosner** [comment]
Hey @jen20! Thanks for getting back to me so quick.

I'm actually building this from scratch, so there shouldn't be any conflicting dependencies or anything of the sort...it's an empty workspace. I opted to download and setup Go 1.5.2 and went with the default /usr/local/go as $GOROOT and that seems to have helped a bit over the yum install.

I now see many more dependencies being downloaded, but now I'm getting an Error 125 as such:

```
github.com/mitchellh/packer (download)
github.com/mitchellh/panicwrap (download)
github.com/mitchellh/prefixedio (download)
github.com/nesv/go-dynect (download)
github.com/packer-community/winrmcp (download)
github.com/packethost/packngo (download)
github.com/pearkes/cloudflare (download)
github.com/pearkes/dnsimple (download)
github.com/pearkes/mailgun (download)
github.com/rackspace/gophercloud (download)
github.com/soniah/dnsmadeeasy (download)
github.com/sthulb/mime (download)
github.com/vmware/govmomi (download)
golang.org/x/net (download)
github.com/xanzy/go-cloudstack (download)
github.com/xanzy/ssh-agent (download)
golang.org/x/crypto (download)
github.com/ziutek/mymysql (download)
golang.org/x/oauth2 (download)
google.golang.org/cloud (download)
google.golang.org/api (download)
github.com/coreos/etcd/Godeps/_workspace/src/github.com/ugorji/go/codec
go build github.com/coreos/etcd/Godeps/_workspace/src/github.com/ugorji/go/codec: /usr/local/go/pkg/tool/linux_amd64/compile: signal: killed
github.com/vmware/govmomi/vim25/types
go build github.com/vmware/govmomi/vim25/types: /usr/local/go/pkg/tool/linux_amd64/compile: signal: killed
github.com/xanzy/go-cloudstack/cloudstack
xargs: go: terminated by signal 9
make: *** [updatedeps] Error 125
```
 
Looks like coreos and vmware is being killed. Any ideas why this might be happening?

**Comment by toddrosner** [comment]
@jen20 the compile problem is solved. I was using a t2.nano which didn't have the required resources; once I upgraded to an instance with multiple vCPUs and more memory, everything worked.

@catsby can you comment on the the first question of my initial comment about controlling the rate of API calls, or how the parallelism flag works? I'm not sure I noticed any difference when using a 0 value for parallelism.

**Comment by toddrosner** [comment]
@jen20 @catsby I haven't heard back, but I thought I would mention that I've upgraded to 0.6.9 and have replaced the NAT config with the new NAT gateway and I'm no longer experiencing any API request rate limits.

I'll keep this open for a few more days to continue testing, if all is good I'll close the issue.

**Comment by catsby** [comment]
Hey @toddrosner sorry for the silence here, have you hit the request limit lately?

**Comment by toddrosner** [comment]
Hey @catsby thanks for getting back. Things have been stable in that way now that I've moved over to the AWS NAT gateways. Previously to that I was building the NATs with instances and ENIs with additional routes over to the new ENI interfaces; this is so that the NATs could auto heal with ASGs and assume the same ENI + route.

I think the process of creating an ENI, attaching it to the instance, then adding a route for some reason introduced some request limit issues...not entirely sure why, but once the AWS NAT gateways were implemented, the request limit issue stopped.

I still think it would be great if it were possible to control the MaxRetries or to better control the parallelism without having to compile terraform and get at the SDK.
